From: olh@suse.de
Subject: patch-2.6.11-rc5-bk4

 -- www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc5-bk3.log	2005-02-28 13:30:42.000000000 +0100
 ++ www.kernel.org/pub/linux/kernel/v2.6/snapshots/patch-2.6.11-rc5-bk4.log	2005-03-01 13:30:05.000000000 +0100
 ChangeSet@1.2065, 2005-02-28 10:33:43-08:00, rddunlap@osdl.org
   [PATCH] rrunner: fix section references
   
   rrunner: references __initdata in a __devinit function; data should
     be __devinitdata;
   
   Error: ./drivers/net/rrunner.o .text refers to 00000000000002b2 R_X86_64_32S      .init.data
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Jes Sorensen <jes@wildopensource.com>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2064, 2005-02-28 10:33:30-08:00, rddunlap@osdl.org
   [PATCH] hp100: fix section references
   
   hp100: references __init code that should be marked as __devinit;
   
   Error: ./drivers/net/hp100.o .text refers to 0000000000000f7f R_X86_64_PC32     .init.text+0x00000000000000b8
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2063, 2005-02-28 10:33:16-08:00, randy.dunlap@verizon.net
   [PATCH] dc395x: fix section references
   
   dc395x scsi driver makes many references to __init functions &
   __initdata that should be __devinit & __devinitdata.
   
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000021ef R_X86_64_PC32     .init.text+0x00000000000000b4
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000220e R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000222e R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000022f4 R_X86_64_PC32     .init.text+0x0000000000000177
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002301 R_X86_64_32S      .init.data
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002307 R_X86_64_32S      .init.data+0x0000000000000004
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000230f R_X86_64_32S      .init.data+0x0000000000000008
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002317 R_X86_64_32S      .init.data+0x000000000000000c
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000231d R_X86_64_32S      .init.data
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000232d R_X86_64_PC32     .init.text+0xfffffffffffffffc
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000237a R_X86_64_PC32     .init.text+0x00000000000000b4
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000238c R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000023ac R_X86_64_PC32     .init.text+0x00000000000000b4
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000023d6 R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000023e6 R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000023fe R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000240b R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002418 R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000242f R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 000000000000243f R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002469 R_X86_64_PC32     .init.text+0x00000000000000b4
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002476 R_X86_64_PC32     .init.text+0x000000000000009c
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002489 R_X86_64_PC32     .init.data+0x0000000000000060
   Error: ./drivers/scsi/dc395x.o .text refers to 0000000000002494 R_X86_64_PC32     .init.text+0x0000000000000177
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000024a3 R_X86_64_32S      .init.data+0x0000000000000078
   Error: ./drivers/scsi/dc395x.o .text refers to 00000000000024ac R_X86_64_PC32     .init.text+0xfffffffffffffffc
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2062, 2005-02-28 10:33:01-08:00, rddunlap@osdl.org
   [PATCH] isdn: use __init for ICCVersion()
   
   ICCVersion() is only used by init code & can be marked __init;
   
   Error: ./drivers/isdn/hisax/icc.o .text refers to 000000000000014a R_X86_64_32S      .init.data
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2061, 2005-02-28 10:32:47-08:00, rddunlap@osdl.org
   [PATCH] sound/oss/opl3as2: fix init section reference
   
   sound/oss/opl3sa2:  calls __init function during probe, which may
   be after init for PNP devices;
   
   Error: ./sound/oss/opl3sa2.o .text refers to 0000000000000204 R_X86_64_PC32     .init.text+0xfffffffffffffffc                                                   Error: ./sound/oss/opl3sa2.o .text refers to 0000000000000210 R_X86_64_PC32     .init.text+0xfffffffffffffffc                                                   Error: ./sound/oss/opl3sa2.o .text refers to 000000000000021c R_X86_64_PC32     .init.text+0xfffffffffffffffc
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2060, 2005-02-28 10:32:33-08:00, rddunlap@osdl.org
   [PATCH] sonicvibes: fix initdata references
   
   sonicvibes:  _devinit function was referencing __initdata (2x),
   which should be __devinitdata;
   
   Error: ./sound/oss/sonicvibes.o .text refers to 0000000000003ca7 R_X86_64_32S      .init.data+0x0000000000000080                                                Error: ./sound/oss/sonicvibes.o .text refers to 00000000000043eb R_X86_64_32S      .init.data+0x0000000000000024                                                Error: ./sound/oss/sonicvibes.o .text refers to 00000000000043f2 R_X86_64_32S      .init.data+0x0000000000000020
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2059, 2005-02-28 10:32:19-08:00, rddunlap@osdl.org
   [PATCH] sound/oss/aedsp16: init/exit section cleanups:
   
   Exit-only function uninit_aedsp16() was marked __init instead of __exit;
   ae_config data was marked __initdata but used during exit;
   several cleanup functions were marked _init but used for init or exit
     cleanups;
   
   Error: ./sound/oss/aedsp16.o .exit.text refers to 0000000000000004 R_X86_64_PC32     .init.data+0x000000000000003b
   Error: ./sound/oss/aedsp16.o .exit.text refers to 000000000000000c R_X86_64_PC32     .init.text+0x00000000000001e4
   Error: ./sound/oss/aedsp16.o .exit.text refers to 0000000000000013 R_X86_64_PC32     .init.text+0x0000000000000034
   Error: ./sound/oss/aedsp16.o .exit.text refers to 0000000000000019 R_X86_64_PC32     .init.data+0x000000000000003f
   Error: ./sound/oss/aedsp16.o .exit.text refers to 0000000000000023 R_X86_64_PC32     .init.text+0x00000000000001b4
   
   Signed-off-by: Randy Dunlap <rddunlap@osdl.org>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2058, 2005-02-28 10:28:19-08:00, axboe@suse.de
   [PATCH] [PATCH] Fix bounced bio and dm panic
   
   Make sure that a bio doesn't contain NULL pages in the front of its vec,
   if a device bounces a bio that doesn't start from 0.
   
   Problem noted by Mark Haverkamp.
   
   Signed-off-by: Jens Axboe <axboe@suse.de>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2057, 2005-02-28 08:36:29-08:00, andrea@suse.de
   [PATCH] Make the new merged pipe writes check for SIGPIPE
   
   The new pipe buffer merging doesn't check for the "no readers" case, so
   with small writes that coalesce, we may not get a timely SIGPIPE/EPIPE
   notification to the writer.
   
   Signed-off-by: Andrea Arcangeli <andrea@suse.de>
   Signed-off-by: Linus Torvalds <torvalds@osdl.org>
 
 ChangeSet@1.2056, 2005-02-28 08:36:14-08:00, torvalds@ppc970.osdl.org
   Make pipe "poll()" take direction of pipe into account.
   
   The pipe code has traditionally not cared about which end-point of the
   pipe you are polling, meaning that if you poll the write-only end of a
   pipe, it will still set the "this pipe is readable" bits if there is
   data to be read on the read side.
   
   That makes no sense, and together with the new bigger buffers breaks
   python-twisted.
   
   Based on debugging/patch by Andrea Arcangeli and testcase from
   Thomas Crhak
 
diff -purN linux-2.6.11-rc5-bk3/Makefile linux-2.6.11-rc5-bk4/Makefile
--- linux-2.6.11-rc5-bk3/Makefile	2005-03-01 14:38:01.005487638 +0100
+++ linux-2.6.11-rc5-bk4/Makefile	2005-03-01 14:38:09.621404573 +0100
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 11
-EXTRAVERSION = -rc5-bk3
+EXTRAVERSION = -rc5-bk4
 NAME=Woozy Numbat
 
 # *DOCUMENTATION*
diff -purN linux-2.6.11-rc5-bk3/drivers/isdn/hisax/icc.c linux-2.6.11-rc5-bk4/drivers/isdn/hisax/icc.c
--- linux-2.6.11-rc5-bk3/drivers/isdn/hisax/icc.c	2005-02-24 05:03:41.000000000 +0100
+++ linux-2.6.11-rc5-bk4/drivers/isdn/hisax/icc.c	2005-03-01 14:38:09.644400995 +0100
@@ -27,7 +27,7 @@
 static char *ICCVer[] __initdata =
 {"2070 A1/A3", "2070 B1", "2070 B2/B3", "2070 V2.4"};
 
-void
+void __init
 ICCVersion(struct IsdnCardState *cs, char *s)
 {
 	int val;
diff -purN linux-2.6.11-rc5-bk3/drivers/isdn/hisax/icc.h linux-2.6.11-rc5-bk4/drivers/isdn/hisax/icc.h
--- linux-2.6.11-rc5-bk3/drivers/isdn/hisax/icc.h	2005-02-24 05:03:10.000000000 +0100
+++ linux-2.6.11-rc5-bk4/drivers/isdn/hisax/icc.h	2005-03-01 14:38:09.645400839 +0100
@@ -65,7 +65,7 @@
 #define ICC_IND_AIL    0xE
 #define ICC_IND_DC     0xF
 
-extern void ICCVersion(struct IsdnCardState *cs, char *s);
+extern void __init ICCVersion(struct IsdnCardState *cs, char *s);
 extern void initicc(struct IsdnCardState *cs);
 extern void icc_interrupt(struct IsdnCardState *cs, u_char val);
 extern void clear_pending_icc_ints(struct IsdnCardState *cs);
diff -purN linux-2.6.11-rc5-bk3/drivers/net/hp100.c linux-2.6.11-rc5-bk4/drivers/net/hp100.c
--- linux-2.6.11-rc5-bk3/drivers/net/hp100.c	2005-02-24 05:02:37.000000000 +0100
+++ linux-2.6.11-rc5-bk4/drivers/net/hp100.c	2005-03-01 14:38:09.651399906 +0100
@@ -306,7 +306,7 @@ static void wait(void)
  * Read board id and convert to string.
  * Effectively same code as decode_eisa_sig
  */
-static __init const char *hp100_read_id(int ioaddr)
+static __devinit const char *hp100_read_id(int ioaddr)
 {
 	int i;
 	static char str[HP100_SIG_LEN];
@@ -429,8 +429,8 @@ struct net_device * __init hp100_probe(i
 }
 #endif
 
-static int __init hp100_probe1(struct net_device *dev, int ioaddr,
-			       u_char bus, struct pci_dev *pci_dev)
+static int __devinit hp100_probe1(struct net_device *dev, int ioaddr,
+				  u_char bus, struct pci_dev *pci_dev)
 {
 	int i;
 	int err = -ENODEV;
diff -purN linux-2.6.11-rc5-bk3/drivers/net/rrunner.c linux-2.6.11-rc5-bk4/drivers/net/rrunner.c
--- linux-2.6.11-rc5-bk3/drivers/net/rrunner.c	2005-02-24 05:02:29.000000000 +0100
+++ linux-2.6.11-rc5-bk4/drivers/net/rrunner.c	2005-03-01 14:38:09.654399439 +0100
@@ -62,7 +62,7 @@ MODULE_AUTHOR("Jes Sorensen <jes@wildope
 MODULE_DESCRIPTION("Essential RoadRunner HIPPI driver");
 MODULE_LICENSE("GPL");
 
-static char version[] __initdata = "rrunner.c: v0.50 11/11/2002  Jes Sorensen (jes@wildopensource.com)\n";
+static char version[] __devinitdata = "rrunner.c: v0.50 11/11/2002  Jes Sorensen (jes@wildopensource.com)\n";
 
 /*
  * Implementation notes:
diff -purN linux-2.6.11-rc5-bk3/drivers/scsi/dc395x.c linux-2.6.11-rc5-bk4/drivers/scsi/dc395x.c
--- linux-2.6.11-rc5-bk3/drivers/scsi/dc395x.c	2005-02-24 05:02:22.000000000 +0100
+++ linux-2.6.11-rc5-bk4/drivers/scsi/dc395x.c	2005-03-01 14:38:09.670396950 +0100
@@ -488,7 +488,7 @@ struct ParameterData {
 	int def;		/* default value */
 	int safe;		/* safe value */
 };
-static struct ParameterData __initdata cfg_data[] = {
+static struct ParameterData __devinitdata cfg_data[] = {
 	{ /* adapter id */
 		CFG_PARAM_UNSET,
 		0,
@@ -573,7 +573,7 @@ MODULE_PARM_DESC(reset_delay, "Reset del
  * set_safe_settings - if the use_safe_settings option is set then
  * set all values to the safe and slow values.
  **/
-static void __init set_safe_settings(void)
+static void __devinit set_safe_settings(void)
 {
 	if (use_safe_settings)
 	{
@@ -592,7 +592,7 @@ static void __init set_safe_settings(voi
  * fix_settings - reset any boot parameters which are out of range
  * back to the default values.
  **/
-static void __init fix_settings(void)
+static void __devinit fix_settings(void)
 {
 	int i;
 
@@ -619,7 +619,7 @@ static void __init fix_settings(void)
  * Mapping from the eeprom delay index value (index into this array)
  * to the the number of actual seconds that the delay should be for.
  */
-static char __initdata eeprom_index_to_delay_map[] = 
+static char __devinitdata eeprom_index_to_delay_map[] = 
 	{ 1, 3, 5, 10, 16, 30, 60, 120 };
 
 
@@ -629,7 +629,7 @@ static char __initdata eeprom_index_to_d
  *
  * @eeprom: The eeprom structure in which we find the delay index to map.
  **/
-static void __init eeprom_index_to_delay(struct NvRamType *eeprom)
+static void __devinit eeprom_index_to_delay(struct NvRamType *eeprom)
 {
 	eeprom->delay_time = eeprom_index_to_delay_map[eeprom->delay_time];
 }
@@ -642,7 +642,7 @@ static void __init eeprom_index_to_delay
  *
  * @delay: The delay, in seconds, to find the eeprom index for.
  **/
-static int __init delay_to_eeprom_index(int delay)
+static int __devinit delay_to_eeprom_index(int delay)
 {
 	u8 idx = 0;
 	while (idx < 7 && eeprom_index_to_delay_map[idx] < delay)
@@ -658,7 +658,7 @@ static int __init delay_to_eeprom_index(
  *
  * @eeprom: The eeprom data to override with command line options.
  **/
-static void __init eeprom_override(struct NvRamType *eeprom)
+static void __devinit eeprom_override(struct NvRamType *eeprom)
 {
 	u8 id;
 
@@ -3931,7 +3931,7 @@ static void dc395x_slave_destroy(struct 
  *
  * @io_port: base I/O address
  **/
-static void __init trms1040_wait_30us(unsigned long io_port)
+static void __devinit trms1040_wait_30us(unsigned long io_port)
 {
 	/* ScsiPortStallExecution(30); wait 30 us */
 	outb(5, io_port + TRM_S1040_GEN_TIMER);
@@ -3948,7 +3948,7 @@ static void __init trms1040_wait_30us(un
  * @cmd:	SB + op code (command) to send
  * @addr:	address to send
  **/
-static void __init trms1040_write_cmd(unsigned long io_port, u8 cmd, u8 addr)
+static void __devinit trms1040_write_cmd(unsigned long io_port, u8 cmd, u8 addr)
 {
 	int i;
 	u8 send_data;
@@ -3993,7 +3993,7 @@ static void __init trms1040_write_cmd(un
  * @addr:	offset into EEPROM
  * @byte:	bytes to write
  **/
-static void __init trms1040_set_data(unsigned long io_port, u8 addr, u8 byte)
+static void __devinit trms1040_set_data(unsigned long io_port, u8 addr, u8 byte)
 {
 	int i;
 	u8 send_data;
@@ -4047,7 +4047,7 @@ static void __init trms1040_set_data(uns
  * @eeprom:	the data to write
  * @io_port:	the base io port
  **/
-static void __init trms1040_write_all(struct NvRamType *eeprom, unsigned long io_port)
+static void __devinit trms1040_write_all(struct NvRamType *eeprom, unsigned long io_port)
 {
 	u8 *b_eeprom = (u8 *)eeprom;
 	u8 addr;
@@ -4087,7 +4087,7 @@ static void __init trms1040_write_all(st
  *
  * Returns the the byte read.
  **/
-static u8 __init trms1040_get_data(unsigned long io_port, u8 addr)
+static u8 __devinit trms1040_get_data(unsigned long io_port, u8 addr)
 {
 	int i;
 	u8 read_byte;
@@ -4125,7 +4125,7 @@ static u8 __init trms1040_get_data(unsig
  * @eeprom:	where to store the data
  * @io_port:	the base io port
  **/
-static void __init trms1040_read_all(struct NvRamType *eeprom, unsigned long io_port)
+static void __devinit trms1040_read_all(struct NvRamType *eeprom, unsigned long io_port)
 {
 	u8 *b_eeprom = (u8 *)eeprom;
 	u8 addr;
@@ -4155,7 +4155,7 @@ static void __init trms1040_read_all(str
  * @eeprom:	caller allocated strcuture to read the eeprom data into
  * @io_port:	io port to read from
  **/
-static void __init check_eeprom(struct NvRamType *eeprom, unsigned long io_port)
+static void __devinit check_eeprom(struct NvRamType *eeprom, unsigned long io_port)
 {
 	u16 *w_eeprom = (u16 *)eeprom;
 	u16 w_addr;
@@ -4225,7 +4225,7 @@ static void __init check_eeprom(struct N
  *
  * @eeprom: The eeprom data strucutre to show details for.
  **/
-static void __init print_eeprom_settings(struct NvRamType *eeprom)
+static void __devinit print_eeprom_settings(struct NvRamType *eeprom)
 {
 	dprintkl(KERN_INFO, "Used settings: AdapterID=%02i, Speed=%i(%02i.%01iMHz), dev_mode=0x%02x\n",
 		eeprom->scsi_id,
@@ -4254,7 +4254,7 @@ static void adapter_sg_tables_free(struc
 /*
  * Allocate SG tables; as we have to pci_map them, an SG list (struct SGentry*)
  * should never cross a page boundary */
-static int __init adapter_sg_tables_alloc(struct AdapterCtlBlk *acb)
+static int __devinit adapter_sg_tables_alloc(struct AdapterCtlBlk *acb)
 {
 	const unsigned mem_needed = (DC395x_MAX_SRB_CNT+1)
 	                            *SEGMENTX_LEN;
@@ -4300,7 +4300,7 @@ static int __init adapter_sg_tables_allo
  *
  * @acb: The adapter to print the information for.
  **/
-static void __init adapter_print_config(struct AdapterCtlBlk *acb)
+static void __devinit adapter_print_config(struct AdapterCtlBlk *acb)
 {
 	u8 bval;
 
@@ -4344,7 +4344,7 @@ static void __init adapter_print_config(
  *
  * @acb: The adapter to initialize.
  **/
-static void __init adapter_init_params(struct AdapterCtlBlk *acb)
+static void __devinit adapter_init_params(struct AdapterCtlBlk *acb)
 {
 	struct NvRamType *eeprom = &acb->eeprom;
 	int i;
@@ -4406,7 +4406,7 @@ static void __init adapter_init_params(s
  *
  * @host: The scsi host instance to fill in the values for.
  **/
-static void __init adapter_init_scsi_host(struct Scsi_Host *host)
+static void __devinit adapter_init_scsi_host(struct Scsi_Host *host)
 {
         struct AdapterCtlBlk *acb = (struct AdapterCtlBlk *)host->hostdata;
 	struct NvRamType *eeprom = &acb->eeprom;
@@ -4447,7 +4447,7 @@ static void __init adapter_init_scsi_hos
  *
  * @acb: The adapter which we are to init.
  **/
-static void __init adapter_init_chip(struct AdapterCtlBlk *acb)
+static void __devinit adapter_init_chip(struct AdapterCtlBlk *acb)
 {
         struct NvRamType *eeprom = &acb->eeprom;
         
@@ -4500,7 +4500,7 @@ static void __init adapter_init_chip(str
  * Returns 0 if the initialization succeeds, any other value on
  * failure.
  **/
-static int __init adapter_init(struct AdapterCtlBlk *acb,
+static int __devinit adapter_init(struct AdapterCtlBlk *acb,
 	unsigned long io_port, u32 io_port_len, unsigned int irq)
 {
 	if (!request_region(io_port, io_port_len, DC395X_NAME)) {
diff -purN linux-2.6.11-rc5-bk3/fs/pipe.c linux-2.6.11-rc5-bk4/fs/pipe.c
--- linux-2.6.11-rc5-bk3/fs/pipe.c	2005-02-24 05:02:56.000000000 +0100
+++ linux-2.6.11-rc5-bk4/fs/pipe.c	2005-03-01 14:38:09.684394772 +0100
@@ -235,6 +235,12 @@ pipe_writev(struct file *filp, const str
 	down(PIPE_SEM(*inode));
 	info = inode->i_pipe;
 
+	if (!PIPE_READERS(*inode)) {
+		send_sig(SIGPIPE, current, 0);
+		ret = -EPIPE;
+		goto out;
+	}
+
 	/* We try to merge small writes */
 	if (info->nrbufs && total_len < PAGE_SIZE) {
 		int lastbuf = (info->curbuf + info->nrbufs - 1) & (PIPE_BUFFERS-1);
@@ -398,13 +404,18 @@ pipe_poll(struct file *filp, poll_table 
 
 	/* Reading only -- no need for acquiring the semaphore.  */
 	nrbufs = info->nrbufs;
-	mask = (nrbufs > 0) ? POLLIN | POLLRDNORM : 0;
-	mask |= (nrbufs < PIPE_BUFFERS) ? POLLOUT | POLLWRNORM : 0;
+	mask = 0;
+	if (filp->f_mode & FMODE_READ) {
+		mask = (nrbufs > 0) ? POLLIN | POLLRDNORM : 0;
+		if (!PIPE_WRITERS(*inode) && filp->f_version != PIPE_WCOUNTER(*inode))
+			mask |= POLLHUP;
+	}
 
-	if (!PIPE_WRITERS(*inode) && filp->f_version != PIPE_WCOUNTER(*inode))
-		mask |= POLLHUP;
-	if (!PIPE_READERS(*inode))
-		mask |= POLLERR;
+	if (filp->f_mode & FMODE_WRITE) {
+		mask |= (nrbufs < PIPE_BUFFERS) ? POLLOUT | POLLWRNORM : 0;
+		if (!PIPE_READERS(*inode))
+			mask |= POLLERR;
+	}
 
 	return mask;
 }
diff -purN linux-2.6.11-rc5-bk3/mm/highmem.c linux-2.6.11-rc5-bk4/mm/highmem.c
--- linux-2.6.11-rc5-bk3/mm/highmem.c	2005-02-24 05:03:04.000000000 +0100
+++ linux-2.6.11-rc5-bk4/mm/highmem.c	2005-03-01 14:38:09.697392750 +0100
@@ -425,7 +425,7 @@ static void __blk_queue_bounce(request_q
 	 * at least one page was bounced, fill in possible non-highmem
 	 * pages
 	 */
-	bio_for_each_segment(from, *bio_orig, i) {
+	__bio_for_each_segment(from, *bio_orig, i, 0) {
 		to = bio_iovec_idx(bio, i);
 		if (!to->bv_page) {
 			to->bv_page = from->bv_page;
diff -purN linux-2.6.11-rc5-bk3/sound/oss/aedsp16.c linux-2.6.11-rc5-bk4/sound/oss/aedsp16.c
--- linux-2.6.11-rc5-bk3/sound/oss/aedsp16.c	2005-02-24 05:02:46.000000000 +0100
+++ linux-2.6.11-rc5-bk4/sound/oss/aedsp16.c	2005-03-01 14:38:09.733387149 +0100
@@ -489,7 +489,7 @@ static struct orVals orDMA[] __initdata 
 	{0x00, 0x00}
 };
 
-static struct aedsp16_info ae_config __initdata = {
+static struct aedsp16_info ae_config = {
 	DEF_AEDSP16_IOB,
 	DEF_AEDSP16_IRQ,
 	DEF_AEDSP16_MRQ,
@@ -1155,7 +1155,7 @@ static int __init init_aedsp16_sb(void)
 	return TRUE;
 }
 
-static void __init uninit_aedsp16_sb(void)
+static void uninit_aedsp16_sb(void)
 {
 	DBG(("uninit_aedsp16_sb: "));
 
@@ -1196,7 +1196,7 @@ static int __init init_aedsp16_mss(void)
 	return TRUE;
 }
 
-static void __init uninit_aedsp16_mss(void)
+static void uninit_aedsp16_mss(void)
 {
 	DBG(("uninit_aedsp16_mss: "));
 
@@ -1237,7 +1237,7 @@ static int __init init_aedsp16_mpu(void)
 	return TRUE;
 }
 
-static void __init uninit_aedsp16_mpu(void)
+static void uninit_aedsp16_mpu(void)
 {
 	DBG(("uninit_aedsp16_mpu: "));
 
@@ -1294,7 +1294,7 @@ static int __init init_aedsp16(void)
 	return initialized;
 }
 
-static void __init uninit_aedsp16(void)
+static void __exit uninit_aedsp16(void)
 {
 	if (ae_config.mss_base != -1)
 		uninit_aedsp16_mss();
diff -purN linux-2.6.11-rc5-bk3/sound/oss/opl3sa2.c linux-2.6.11-rc5-bk4/sound/oss/opl3sa2.c
--- linux-2.6.11-rc5-bk3/sound/oss/opl3sa2.c	2005-02-24 05:02:16.000000000 +0100
+++ linux-2.6.11-rc5-bk4/sound/oss/opl3sa2.c	2005-03-01 14:38:09.735386838 +0100
@@ -711,7 +711,7 @@ static void __init attach_opl3sa2_mixer(
 }
 
 
-static void __init opl3sa2_clear_slots(struct address_info* hw_config)
+static void opl3sa2_clear_slots(struct address_info* hw_config)
 {
 	int i;
 
diff -purN linux-2.6.11-rc5-bk3/sound/oss/sonicvibes.c linux-2.6.11-rc5-bk4/sound/oss/sonicvibes.c
--- linux-2.6.11-rc5-bk3/sound/oss/sonicvibes.c	2005-02-24 05:02:38.000000000 +0100
+++ linux-2.6.11-rc5-bk4/sound/oss/sonicvibes.c	2005-03-01 14:38:09.740386060 +0100
@@ -2470,7 +2470,7 @@ MODULE_LICENSE("GPL");
 static struct initvol {
 	int mixch;
 	int vol;
-} initvol[] __initdata = {
+} initvol[] __devinitdata = {
 	{ SOUND_MIXER_WRITE_RECLEV, 0x4040 },
 	{ SOUND_MIXER_WRITE_LINE1, 0x4040 },
 	{ SOUND_MIXER_WRITE_CD, 0x4040 },
@@ -2487,7 +2487,7 @@ static struct initvol {
 
 static int __devinit sv_probe(struct pci_dev *pcidev, const struct pci_device_id *pciid)
 {
-	static char __initdata sv_ddma_name[] = "S3 Inc. SonicVibes DDMA Controller";
+	static char __devinitdata sv_ddma_name[] = "S3 Inc. SonicVibes DDMA Controller";
        	struct sv_state *s;
 	mm_segment_t fs;
 	int i, val, ret;
